### PR TITLE
ENH hide all symbols

### PR DESCRIPTION
**Pull request by beckermr**
This PR tries to hide all of the cfitsio symbols from the python extension so that the linker does not get confused.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def build_extensions(self):
             # (-fPIC), since we use the python compiler flags
 
             link_objects = glob.glob(
-                os.path.join(self.cfitsio_build_dir, '*.a'))
+                os.path.join(self.cfitsio_build_dir, '*.o'))
```
Comment (beckermr) [inline setup.py:151]: This is a big change. It is needed to hide the symbols. We now link against the .o files as if we were bundle the C code directly, not just bundling a static lib.

File: setup.py [hunk 2]
```diff
@@ -292,7 +299,11 @@ def check_system_cfitsio_objects(self, obj_name):
 
 sources = ["fitsio/fitsio_pywrap.c"]
 
-ext = Extension("fitsio._fitsio_wrap", sources, include_dirs=['numpy'])
+ext = Extension(
+    "fitsio._fitsio_wrap",
+    sources,
+    include_dirs=['numpy'],
+)
```
Comment (beckermr) [inline setup.py:306]:
```suggestion
ext = Extension("fitsio._fitsio_wrap", sources, include_dirs=['numpy'])
```

File: setup.py [hunk 3]
```diff
@@ -247,7 +247,9 @@ def copy_update(dir1, dir2):
         args = ''
         if CC is not None:
             args += ' CC="%s"' % ' '.join(CC[:1])
-            args += ' CFLAGS="%s"' % ' '.join(CC[1:])
+            args += ' CFLAGS="%s -fvisibility=hidden"' % ' '.join(CC[1:])
```
Comment (esheldon) [inline setup.py:250]: Is this needed?  Was using the .o files good enough?

Comment (beckermr) [reply setup.py:250]: We didn't test that with Klaus. I will rebuild and check the libs locally. 

However given the docs on gcc, I think we want both.

Comment (beckermr) [reply setup.py:250]: Actually I think we need it. Before I added the static qualifier to some helper functions in the extension C code, those were appearing as exported symbols in the final lib. I don't think the entire cfitsio code base has static applied.

Comment (beckermr) [reply setup.py:250]: Yup. The symbols are visible without those flags. We did not test explicitly if the thing you are asking about.

**Review by esheldon — APPROVED** [review]

**Pull request merged**